core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,425 @@
+   * before any other data goes in. The purpose of padding is to put the input
+   * data at proper position.
+   */
+  var padding: Byte = '0'
+  var closed: Boolean = false
+  var key: Array[Byte] = keyVal.clone()
+  var initIV: Array[Byte] = ivVal.clone()
+  var iv: Array[Byte] = ivVal.clone()
+  var isReadableByteChannel: Boolean = in.isInstanceOf[ReadableByteChannel]
+
+  /** DirectBuffer pool */
+  var bufferPool: Queue[ByteBuffer] = new ConcurrentLinkedQueue[ByteBuffer]()
+  /** Decryptor pool */
+  var decryptorPool: Queue[Decryptor] = new ConcurrentLinkedQueue[Decryptor]()
+
+  lazy val tmpBuf: Array[Byte] = new Array[Byte](bufferSize)
+  var decryptor: Decryptor = getDecryptor()
+  CryptoStreamUtils.checkCodec(codec)
+  resetStreamOffset(streamOffset)
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      bufferSize: Integer,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, bufferSize, key, iv, 0, true)
+  }
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, CryptoStreamUtils.getBufferSize, key, iv)
+  }
+
+  def getWrappedStream(): InputStream = in
+
+  /**
+   * Decryption is buffer based.
+   * If there is data in [[outBuffer]], then read it out of this buffer.
+   * If there is no data in [[outBuffer]], then read more from the
+   * underlying stream and do the decryption.
+   * @param b the buffer into which the decrypted data is read.
+   * @param off the buffer offset.
+   * @param len the maximum number of decrypted data bytes to read.
+   * @return int the total number of decrypted data bytes read into the buffer.
+   * @throws IOException
+   */
+  override def read(b: Array[Byte], off: Int, len: Int): Int = {
+    checkStream()
+    Preconditions.checkNotNull(b)
+    Preconditions.checkArgument(!(off < 0 || len < 0 || len > b.length - off))
+
+    if (len == 0) {
+      0
+    } else {
+      val remaining = outBuffer.remaining()
+      if (remaining > 0) {
+        val n = Math.min(len, remaining)
+        outBuffer.get(b, off, n)
+        n
+      } else {
+        var n = 0
+        if (usingByteBufferReadInitialized == false) {
+          usingByteBufferReadInitialized = true
+          if (isReadableByteChannel) {
+            try {
+              n = (in.asInstanceOf[ReadableByteChannel]).read(inBuffer)
+              usingByteBufferRead = true
+            } catch {
+              case e: UnsupportedOperationException =>
+                usingByteBufferRead = false
+            }
+          } else {
+            usingByteBufferRead = false
+          }
+          if (!usingByteBufferRead) {
+            n = readFromUnderlyingStream(inBuffer)
+          }
+        } else {
+          if (usingByteBufferRead) {
+            n = (in.asInstanceOf[ReadableByteChannel]).read(inBuffer)
+          } else {
+            n = readFromUnderlyingStream(inBuffer)
+          }
+        }
+        if (n <= 0) {
```

Comment:
I'd prefer:

    if (n > 0) {
      // do the stuff in the else
    }
    n